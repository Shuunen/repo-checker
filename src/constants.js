import { join } from 'path'

export const home = process.env.HOME
export const dataFileName = 'repo-checker.config.js'
export const repoCheckerPath = join(__dirname, '..')
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export const dataDefaults = {
  auto_merge: true,
  ban_sass: true,
  dev_deps_only: true,
  license: 'GPL-3.0',
  max_size_ko: 10,
  user_mail: 'romain.racamier@gmail.com',
  user_name: 'Romain Racamier-Lafon',
}
