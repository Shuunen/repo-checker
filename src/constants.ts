/* c8 ignore next 4 */
// biome-ignore lint/correctness/noNodejsModules: we are in a nodejs environment
import path from 'node:path'

export const home = process.env.HOME ?? '' // when does HOME is not defined ?
/**
 * The name of the file that contains the configuration for repo-checker
 */
export const dataFileName = '.repo-checker.json'
// eslint-disable-next-line unicorn/prefer-module
export const repoCheckerPath = path.join(__dirname, '..')
export const templatePath = path.join(repoCheckerPath, 'templates')

/**
 * The project data/configuration for repo-checker
 */
// eslint-disable-next-line no-restricted-syntax
export class ProjectData {
  /**
   * Renovate auto-merge feature
   */
  public canAutoMergeDeps = true

  public hasMainBranch = false

  public hasTaskPrefix = false

  public isModule = false

  public isPublishedPackage = false

  public isQuiet = false

  public isUsingC8 = false

  public isUsingDependencyCruiser = false

  public isUsingEslint = false

  public isUsingNyc = false

  public isUsingShuutils = false

  public isUsingTailwind = false

  public isUsingTypescript = false

  public isUsingV8 = false

  public isUsingVue = false

  public isWebPublished = false

  public license = 'GPL-3.0'

  public maxSizeKo = 1

  public packageName = ''

  public repoId = 'a-great-repo'

  public shouldAvoidSass = true

  public userId = 'Shuunen'

  public userIdLowercase = 'shuunen'

  public userMail = 'romain.racamier@gmail.com'

  public userName = 'Romain Racamier-Lafon'

  public webUrl = 'https://my-website.com'

  public willGenerateReport = true

  /**
   * Create a new project data
   * @param data the data to use
   */
  public constructor(data: Readonly<Partial<ProjectData>> = {}) {
    Object.assign(this, data)
  }
}

export const dataDefaults = new ProjectData()
