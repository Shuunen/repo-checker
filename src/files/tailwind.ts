import { FileBase } from '../file'

export class TailwindFile extends FileBase {
  public async start (): Promise<void> {
    if (!this.data.isUsingTailwind) return
    const { fileName, hasTsFile } = await this.detectContext()
    await this.checkFileExists(fileName)
    await this.inspectFile(fileName)
    if (!hasTsFile) {
      const hasTypes = this.shouldContains('type definitions', /@type/u, 1, true, 'like : /** @type {import(\'tailwindcss\').Config} */', true)
      if (!hasTypes && this.canFix) this.fileContent = `/** @type {import('tailwindcss').Config} */\n${this.fileContent}`
    }
    this.shouldContains('a content (previously named purge) option', /content/u, 1, false, 'like : content: [\'./src/**/*.{vue,js,ts,jsx,tsx}\']')
  }

  private async detectContext (): Promise<{ fileName: string; hasTsFile: boolean }> {
    const hasJsFile = await this.fileExists('tailwind.config.js')
    const hasTsFile = await this.fileExists('tailwind.config.ts')
    const hasMjsFile = await this.fileExists('tailwind.config.mjs')
    /* c8 ignore next 2 */
    // eslint-disable-next-line no-nested-ternary
    const fileName = `tailwind.config.${hasMjsFile ? 'mjs' : hasTsFile ? 'ts' : hasJsFile ? 'js' : 'cjs'}`
    return { fileName, hasTsFile }
  }
}
