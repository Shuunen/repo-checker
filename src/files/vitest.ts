import { FileBase } from '../file.ts'

// eslint-disable-next-line no-export
export class VitestFile extends FileBase {
  // eslint-disable-next-line jsdoc/require-jsdoc
  // eslint-disable-next-line max-lines-per-function
  public async start() {
    if (!this.data.isUsingVitest) return
    const hasMtsFile = await this.fileExists('vitest.config.mts')
    const hasJsFile = await this.fileExists('vitest.config.js')
    const hasTsFile = await this.fileExists('vitest.config.ts')
    const hasMjsFile = await this.fileExists('vitest.config.mjs')
    if (hasMtsFile) await this.inspectFile('vitest.config.mts')
    /* c8 ignore next 3 */
    if (hasJsFile) await this.inspectFile('vitest.config.js')
    if (hasTsFile) await this.inspectFile('vitest.config.ts')
    if (hasMjsFile) await this.inspectFile('vitest.config.mjs')
    this.couldContains('a pool thread setting', /pool: 'threads'/u)
  }
}
