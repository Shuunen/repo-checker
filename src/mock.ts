/* eslint-disable jsdoc/require-jsdoc */
import path from 'node:path'
import { clone, sleep } from 'shuutils'
import type { Indicators } from './check.ts'
import type { FileBase } from './file.ts'
import { join } from './utils.ts'

const fileBaseKeysToDelete: (keyof FileBase)[] = ['fileExists', 'checkFileExists', 'folderPath', 'inspectFile', 'updateFile', 'initFile']
const indicatorsKeys: ReadonlyArray<keyof Indicators> = ['failed', 'passed', 'warnings'] as const

function cleanStringForSnap(input: string) {
  return (
    input
      // replace repo-check-1-40 by repo-check-x-yy
      .replace(/repo-check-\d-\d+/gu, 'repo-check-x-yy')
      // replace package-json-main-file-size-63ko by package-json-main-file-size-XYZko
      .replace(/package-json-main-file-size-\d+ko/gu, 'package-json-main-file-size-XYZko')
  )
}

// eslint-disable-next-line no-restricted-syntax
export function cleanUnknownValueForSnap<Type>(input: Type): Type {
  /* c8 ignore next 8 */
  /* eslint-disable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
  if (typeof input === 'string') return cleanStringForSnap(input) as Type
  if (Array.isArray(input)) return input.map(item => cleanUnknownValueForSnap(item)) as Type
  if (typeof input === 'object') return JSON.parse(cleanStringForSnap(JSON.stringify(input))) as Type
  // eslint-disable-next-line no-restricted-syntax
  throw new Error(`cleanInstanceValueForSnap: unknown type ${typeof input}`)
  /* eslint-enable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
}

export async function promiseValue<Type>(value: Type) {
  await sleep(1)
  return value
}

export async function promiseTrue() {
  await sleep(1)
  return true
}

export async function promiseFalse() {
  await sleep(1)
  return false
}

export async function promiseVoid() {
  await sleep(1)
}

// eslint-disable-next-line unicorn/prefer-module
export const sourceFolder = path.dirname(__filename)

export const vueProjectFolder = join(sourceFolder, 'mocks', 'vueProject')

export const tsProjectFolder = join(sourceFolder, 'mocks', 'tsProject')

export const mocksProjectsFolder = join(sourceFolder, 'mocks')

export function cleanIndicatorsForSnap(record?: Readonly<Record<string, unknown>>) {
  if (!record) return record
  const clean = clone(record)
  for (const key of indicatorsKeys) if (key in clean) clean[key] = cleanUnknownValueForSnap(clean[key])
  return clean
}

// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
export function cleanInstanceForSnap(instance: Readonly<FileBase>, ...bonusKeys: readonly string[]) {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const clean = clone<Partial<FileBase>>(instance) as Record<string, unknown>
  const keys = [...fileBaseKeysToDelete, ...bonusKeys] // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
  for (const key of keys) if (key in clean) delete clean[key]
  return cleanIndicatorsForSnap(clean)
}
