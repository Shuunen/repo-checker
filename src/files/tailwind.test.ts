import { expect, it } from 'vitest'
import { ProjectData } from '../constants'
import { cleanInstanceForSnap, promiseValue, vueProjectFolder } from '../mock'
import { TailwindFile } from './tailwind'

it('tailwind : can detect valid config', async () => {
  const instance = new TailwindFile(vueProjectFolder, new ProjectData({ isQuiet: true, isUsingTailwind: true }), true)
  instance.initFile = promiseValue // prevent file creation
  await instance.start()
  instance.fileName = '' // prevent file update
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})
