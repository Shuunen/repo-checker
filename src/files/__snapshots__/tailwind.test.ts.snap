// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tailwind : can detect valid config 1`] = `
{
  "canFix": true,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": true,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": false,
    "isUsingShuutils": false,
    "isUsingTailwind": true,
    "isUsingTypescript": false,
    "isUsingV8": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "/** @type {import('tailwindcss').Config} */
// eslint-disable-next-line unicorn/prefer-module
module.exports = {
  content: ['public/*.html', './src/**/*.{vue,js,ts,jsx,tsx}'],
  safelist: [
    {
      pattern: /(?:bg|from|text|to)-\\w+-(?:700|900)/u,
    },
  ],
  theme: {
    extend: {},
  },
}
",
  "fileName": "",
  "originalFileContent": "// eslint-disable-next-line unicorn/prefer-module
module.exports = {
  content: ['public/*.html', './src/**/*.{vue,js,ts,jsx,tsx}'],
  safelist: [
    {
      pattern: /(?:bg|from|text|to)-\\w+-(?:700|900)/u,
    },
  ],
  theme: {
    extend: {},
  },
}
",
  "passed": [
    "has-a-tailwind-config-js-file",
    "tailwind-config-js-has-a-content-previously-named-purge-option",
  ],
  "warnings": [],
}
`;
